File: src_backup/main/java/com/myudog/myulib/internal/generator/ConeGenerator.java
package com.myudog.myulib.internal.generator;
import com.myudog.myulib.api.Shapes;
import com.myudog.myulib.api.core.shape.IShape;
import net.minecraft.world.phys.Vec3;import java.util.ArrayList;
import java.util.List;
public final class ConeGenerator implements IShape {
    @Override
    public List<Vec3> getOutlinePoints(Vec3 size, double density) {
        List<Vec3> points = new ArrayList<>();
        double radiusX = size.x;
        double radiusZ = size.z;
        double height = size.y;
        int circleCount = Math.max(8, (int) Math.round(2.0 * Math.PI * Math.max(radiusX, radiusZ) * Math.max(1.0, density)));
        for (int i = 0; i < circleCount; i++) {
            double angle = 2.0 * Math.PI * i / circleCount;
            points.add(new Vec3(radiusX * Math.cos(angle), 0.0, radiusZ * Math.sin(angle)));
        }
        for (int i = 0; i < 8; i++) {
            double angle = 2.0 * Math.PI * i / 8.0;
            Vec3 basePos = new Vec3(radiusX * Math.cos(angle), 0.0, radiusZ * Math.sin(angle));
            Vec3 tipPos = new Vec3(0.0, height, 0.0);
            int lineDensity = Math.max(1, (int) Math.round(height * Math.max(1.0, density)));
            for (int j = 0; j <= lineDensity; j++) {
                double t = j / (double) lineDensity;
                points.add(new Vec3(
                        basePos.x + (tipPos.x - basePos.x) * t,
                        basePos.y + (tipPos.y - basePos.y) * t,
                        basePos.z + (tipPos.z - basePos.z) * t
                ));
            }
        }
        return points;
    }
    @Override
    public List<Vec3> getSolidPoints(Vec3 size, double density) {
        List<Vec3> points = new ArrayList<>();
        double hStep = 1.0 / Math.max(0.1, density);
        for (double h = 0.0; h <= size.y; h += hStep) {
            double ratio = size.y <= 0.0 ? 0.0 : 1.0 - (h / size.y);
            Vec3 currentRadius = new Vec3(size.x * ratio, 0.0, size.z * ratio);
            for (Vec3 point : Shapes.CIRCLE.getSolidPoints(currentRadius, density)) {
                points.add(new Vec3(point.x, h, point.z));
            }
        }
        return points;
    }
}

File: src/main/kotlin/com/myudog/myulib/internal/generator/ConeGenerator.kt
package com.myudog.myulib.internal.generator

import com.myudog.myulib.api.shape.IShape
import net.minecraft.util.math.Vec3d
import kotlin.math.*

internal class ConeGenerator : IShape {
    override fun getOutlinePoints(size: Vec3d, density: Double): List<Vec3d> {
        val points = mutableListOf<Vec3d>()
        val radiusX = size.x
        val radiusZ = size.z
        val height = size.y

        // 1. 底圓邊框
        val circleCount = (2 * PI * max(radiusX, radiusZ) * density).toInt()
        for (i in 0 until circleCount) {
            val angle = 2 * PI * i / circleCount
            points.add(Vec3d(radiusX * cos(angle), 0.0, radiusZ * sin(angle)))
        }

        // 2. 側面斜線 (從頂點到底圓選 8 條斜線作為骨架)
        for (i in 0 until 8) {
            val angle = 2 * PI * i / 8
            val basePos = Vec3d(radiusX * cos(angle), 0.0, radiusZ * sin(angle))
            val tipPos = Vec3d(0.0, height, 0.0)
            val lineDensity = (height * density).toInt()
            for (j in 0..lineDensity) {
                points.add(basePos.lerp(tipPos, j.toDouble() / lineDensity))
            }
        }
        return points
    }

    override fun getSolidPoints(size: Vec3d, density: Double): List<Vec3d> {
        val points = mutableListOf<Vec3d>()
        val hStep = 1.0 / density
        var h = 0.0
        while (h <= size.y) {
            // 隨高度增加，圓盤半徑線性遞減 (r = R * (1 - h/H))
            val ratio = 1.0 - (h / size.y)
            val currentRadius = Vec3d(size.x * ratio, 0.0, size.z * ratio)

            // 複用 Circle 的實心計算
            com.myudog.myulib.api.Shapes.CIRCLE.getSolidPoints(currentRadius, density).forEach {
                points.add(Vec3d(it.x, h, it.z))
            }
            h += hStep
        }
        return points
    }
}